integration-tests/test/test_wallets.py

```diff
@@ -0,0 +1,96 @@
+    docker_network_with_started_bootstrap,
+)
+from .common import TestingContext
+from . import conftest
+from .wait import wait_for_log_match
+
+
+ALICE_KEY = conftest.KeyPair(private_key='b2527b00340a83e302beae2a8daf6d654e8e57541acfa261cc1b5635eb16aa15', public_key='12a764e25517b99c35db937115f480a5067f6a809dd0c1f382ad2a9212a012c3')
+BOB_KEY = conftest.KeyPair(private_key='9a801debae8bb97fe54c99389cafa576c60612503348578125b65ab182ff5850', public_key='155244935e80a72b4f3e83fb2fcfbb8d07c5bfdbc91e3b6c96371583139acad8')
+
+def deploy_contract_with_substitution(node: Node, substitute_rules: str, contract_name: str, private_key: Optional[str]=None) -> str:
+    local_contract_file_path = os.path.join('resources/wallets', contract_name)
+    node_deploy_file_path = os.path.join(node.local_deploy_dir, contract_name)
+    shutil.copyfile(local_contract_file_path, node_deploy_file_path)
+    container_contract_file_path = '{}/{}'.format(node.remote_deploy_dir, contract_name)
+    node.shell_out(
+        'sed',
+        '-i',
+        '-e', substitute_rules,
+        container_contract_file_path,
+    )
+    node.deploy(container_contract_file_path, private_key=private_key)
+    block_hash = node.propose()
+    os.remove(node_deploy_file_path)
+    return block_hash
+
+def create_genesis_vault(context: TestingContext, node: Node, rev_addr: str, init_vault_value: int) -> None:
+    log_marker = random_string(context, 10)
+    create_vault_success_pattern = re.compile("{} Genesis vault created!".format(log_marker))
+    deploy_contract_with_substitution(node, "s/%REV_ADDR/{}/g; s/%INIT_VAULT_VALUE/{}/g; s/%LOG_MARKER/{}/g".format(rev_addr, init_vault_value, log_marker), "create_genesis_vault.rho")
+    wait_for_log_match(context, node, create_vault_success_pattern)
+
+def transfer_funds(context: TestingContext, node: Node, from_rev_addr: str, to_rev_addr: str, amount: int, private_key: str) -> bool:
+    log_marker = random_string(context, 10)
+    transfer_funds_result_pattern= re.compile('"{} (Successfully|Failing) reason: (?P<reason>[a-zA-Z0-9]*)"'.format(log_marker))
+    deploy_contract_with_substitution(node, "s/%FROM/{}/g; s/%TO/{}/g; s/%AMOUNT/{}/g; s/%LOG_MARKER/{}/g".format(from_rev_addr, to_rev_addr, amount, log_marker), "transfer_funds.rho", private_key)
+    wait_for_log_match(context, node, transfer_funds_result_pattern)
+    transfer_result_match = transfer_funds_result_pattern.search(node.logs())
+    reason = transfer_result_match.group('reason')  # type:ignore
+    return reason == 'Nil'
+
+def check_balance(context: TestingContext, node: Node, rev_addr: str) -> int:
+    log_marker = random_string(context, 10)
+    check_balance_pattern = re.compile('"{} Vault (?P<rev_addr>[a-zA-Z0-9]*) balance is (?P<balance>[0-9]*)"'.format(log_marker))
+    deploy_contract_with_substitution(node, "s/%REV_ADDR/{}/g; s/%LOG_MARKER/{}/g".format(rev_addr, log_marker), "check_balance.rho")
+    wait_for_log_match(context, node, check_balance_pattern)
```

Comment:
I think some of this should be extracted into a separate function `wait_for_balance()`.  It can return the balance if that's possible, avoiding doing the regex search twice.

Reply:
agree. See https://github.com/rchain/rchain/pull/2399#discussion_r280415837.